### PR TITLE
resolve: Refactor how the prelude is handled

### PR DESCRIPTION
**Pull request by jseyfried**
This PR refactors how the prelude is handled in `resolve`.

Instead of importing names from the prelude into each module's `resolutions`, this PR adds a new field `prelude: RefCell<Option<Module>>` to `ModuleS` that is set during import resolution but used only when resolving in a lexical scope (i.e. the scope of an initial segment of a relative path).

r? @nikomatsakis

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @petrochenkov

**Comment by petrochenkov** [comment]
Do you plan to change the prelude injection code in libsyntax? I'd suppose that with these changes prelude can be injected once in the crate root. (Is it ever useful to have different preludes used in different modules?)

**Comment by jseyfried** [comment]
I wasn't planning on refactoring the prelude injection code in `libsyntax` in the near future (this PR is good enough groundwork for the new import semantics), but I think it would be an improvement.

If we decide to stabilize something allowing for different preludes for different modules, we could always make the the field `prelude` per-module again pretty easily.

**Comment by nikomatsakis** [comment]
Can you elaborate a bit more on the larger picture you have in mind for this change? Sorry, there's been a lot in flight lately -- especially around resolve, huzzah! -- and I feel like I'm losing track.

**Comment by jseyfried** [comment]
> Can you elaborate a bit more on the larger picture you have in mind for this change?

Sure -- this is a pure refactoring. The primary motivation is to reflect the interpretation of the prelude we discussed in #31908 and to simplify the core import resolution algorithm by removing `DefModifiers::PRELUDE` bindings.

The last commit lays the groundwork for item-like private imports by making a distinction between resolving in a lexical scope (i.e. allowing private imports and the prelude) and allowing private imports but not the prelude. Right now, we never need to allow private imports and not the prelude, but eventually ...
```rust
struct Bar;
mod foo {
    use Bar;
    use foo::Bar; // ... we will want this to resolve (a private import),
    use foo::Ok; // but not this (a name from the prelude).
}
```

File: src/librustc_resolve/build_reduced_graph.rs [hunk 1]
```diff
@@ -670,15 +664,16 @@ impl<'a, 'b:'a, 'tcx:'b> GraphBuilder<'a, 'b, 'tcx> {
                 module_.increment_outstanding_references_for(target, ValueNS, is_public);
                 module_.increment_outstanding_references_for(target, TypeNS, is_public);
             }
-            GlobImport => {
+            GlobImport if !is_prelude => {
                 // Set the glob flag. This tells us that we don't know the
                 // module's exports ahead of time.
                 module_.inc_glob_count(is_public)
             }
+            _ => {}
```
Comment (nikomatsakis) [inline src/librustc_resolve/build_reduced_graph.rs]: I would suggest avoiding `_` but rather keeping the match exhaustive, with a comment explaining why this case requires no action. This has two advantages:

- Helps the reader know why the prelude case is different.
- Gives you errors if you add new enum variants.

**Comment by nikomatsakis** [comment]
Looks good, r=me modulo the match.

File: src/librustc_resolve/build_reduced_graph.rs (hunk 1 above)
Comment (jseyfried) [inline src/librustc_resolve/build_reduced_graph.rs]: Good point

**Pull request merged**